// aws-cdk/lib/cli/cdk-toolkit.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __addDisposableResource = (this && this.__addDisposableResource) || function (env, value, async) {
    if (value !== null && value !== void 0) {
        if (typeof value !== "object" && typeof value !== "function") throw new TypeError("Object expected.");
        var dispose, inner;
        if (async) {
            if (!Symbol.asyncDispose) throw new TypeError("Symbol.asyncDispose is not defined.");
            dispose = value[Symbol.asyncDispose];
        }
        if (dispose === void 0) {
            if (!Symbol.dispose) throw new TypeError("Symbol.dispose is not defined.");
            dispose = value[Symbol.dispose];
            if (async) inner = dispose;
        }
        if (typeof dispose !== "function") throw new TypeError("Object not disposable.");
        if (inner) dispose = function() { try { inner.call(this); } catch (e) { return Promise.reject(e); } };
        env.stack.push({ value: value, dispose: dispose, async: async });
    }
    else if (async) {
        env.stack.push({ async: true });
    }
    return value;
};
var __disposeResources = (this && this.__disposeResources) || (function (SuppressedError) {
    return function (env) {
        function fail(e) {
            env.error = env.hasError ? new SuppressedError(e, env.error, "An error was suppressed during disposal.") : e;
            env.hasError = true;
        }
        var r, s = 0;
        function next() {
            while (r = env.stack.pop()) {
                try {
                    if (!r.async && s === 1) return s = 0, env.stack.push(r), Promise.resolve().then(next);
                    if (r.dispose) {
                        var result = r.dispose.call(r.value);
                        if (r.async) return s |= 2, Promise.resolve(result).then(next, function(e) { fail(e); return next(); });
                    }
                    else s |= 1;
                }
                catch (e) {
                    fail(e);
                }
            }
            if (s === 1) return env.hasError ? Promise.reject(env.error) : Promise.resolve();
            if (env.hasError) throw env.error;
        }
        return next();
    };
})(typeof SuppressedError === "function" ? SuppressedError : function (error, suppressed, message) {
    var e = new Error(message);
    return e.name = "SuppressedError", e.error = error, e.suppressed = suppressed, e;
});
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.CdkToolkit = exports.AssetBuildTime = void 0;
exports.displayFlagsMessage = displayFlagsMessage;
const node_crypto_1 = require("node:crypto");
const path = __importStar(require("node:path"));
const node_util_1 = require("node:util");
const cxapi = __importStar(require("@aws-cdk/cloud-assembly-api"));
const cloud_assembly_schema_1 = require("@aws-cdk/cloud-assembly-schema");
const toolkit_lib_1 = require("@aws-cdk/toolkit-lib");
const chalk_1 = __importDefault(require("chalk"));
const chokidar = __importStar(require("chokidar"));
const handler_js_1 = require("chokidar/handler.js");
const fs = __importStar(require("fs-extra"));
const io_host_1 = require("./io-host");
const user_configuration_1 = require("./user-configuration");
const api_private_1 = require("../../lib/api-private");
const api_1 = require("../api");
const bootstrap_1 = require("../api/bootstrap");
const cloud_assembly_1 = require("../api/cloud-assembly");
const deploy_private_1 = require("../api/deploy-private");
const refactor_1 = require("../api/refactor");
const deploy_1 = require("../commands/deploy");
const list_stacks_1 = require("../commands/list-stacks");
const migrate_1 = require("../commands/migrate");
const cxapp_1 = require("../cxapp");
const util_1 = require("../util");
const collect_telemetry_1 = require("./telemetry/collect-telemetry");
const error_1 = require("./telemetry/error");
const messages_1 = require("./telemetry/messages");
const operations_1 = require("../commands/flags/operations");
// Must use a require() otherwise esbuild complains about calling a namespace
// eslint-disable-next-line @typescript-eslint/no-require-imports,@typescript-eslint/consistent-type-imports
const pLimit = require('p-limit');
/**
 * File events that we care about from chokidar.
 * In chokidar v4, EventName includes additional events like 'error', 'raw', 'ready', 'all'
 * that we need to filter out in the 'all' handler.
 */
const FILE_EVENTS = [handler_js_1.EVENTS.ADD, handler_js_1.EVENTS.ADD_DIR, handler_js_1.EVENTS.CHANGE, handler_js_1.EVENTS.UNLINK, handler_js_1.EVENTS.UNLINK_DIR];
/**
 * Type guard to check if an event is a file event we should process.
 */
function isFileEvent(event) {
    return FILE_EVENTS.includes(event);
}
/**
 * When to build assets
 */
var AssetBuildTime;
(function (AssetBuildTime) {
    /**
     * Build all assets before deploying the first stack
     *
     * This is intended for expensive Docker image builds; so that if the Docker image build
     * fails, no stacks are unnecessarily deployed (with the attendant wait time).
     */
    AssetBuildTime["ALL_BEFORE_DEPLOY"] = "all-before-deploy";
    /**
     * Build assets just-in-time, before publishing
     */
    AssetBuildTime["JUST_IN_TIME"] = "just-in-time";
})(AssetBuildTime || (exports.AssetBuildTime = AssetBuildTime = {}));
/**
 * Custom implementation of the public Toolkit to integrate with the legacy CdkToolkit
 *
 * This overwrites how an sdkProvider is acquired
 * in favor of the one provided directly to CdkToolkit.
 */
class InternalToolkit extends toolkit_lib_1.Toolkit {
    _sdkProvider;
    constructor(sdkProvider, options) {
        super(options);
        this._sdkProvider = sdkProvider;
    }
    /**
     * Access to the AWS SDK
     * @internal
     */
    async sdkProvider(_action) {
        return this._sdkProvider;
    }
}
/**
 * Toolkit logic
 *
 * The toolkit runs the `cloudExecutable` to obtain a cloud assembly and
 * deploys applies them to `cloudFormation`.
 */
class CdkToolkit {
    props;
    ioHost;
    toolkitStackName;
    toolkit;
    constructor(props) {
        this.props = props;
        this.ioHost = props.ioHost ?? io_host_1.CliIoHost.instance();
        this.toolkitStackName = props.toolkitStackName ?? api_1.DEFAULT_TOOLKIT_STACK_NAME;
        // We don't have the toolkit check unstable features. Instead, the error
        // messages the CLI give are slightly different, and so they need to be checked
        // separately. Instead, the Toolkit we use has all unstable features automatically
        // enabled.
        const iPromiseUnstablenessIsCheckedByTheCli = {
            'publish-assets': true,
            'diagnose': true,
            'flags': true,
            'orphan': true,
            'refactor': true,
            'validate': true,
        };
        this.toolkit = new InternalToolkit(props.sdkProvider, {
            assemblyFailureAt: this.validateMetadataFailAt(),
            color: true,
            emojis: true,
            ioHost: this.ioHost,
            toolkitStackName: this.toolkitStackName,
            unstableFeatures: Object.keys(iPromiseUnstablenessIsCheckedByTheCli),
        });
    }
    async metadata(stackName, json) {
        const stacks = await this.selectSingleStackByName(stackName);
        await printSerializedObject(this.ioHost.asIoHelper(), stacks.firstStack.metadata ?? {}, json);
    }
    async acknowledge(noticeId) {
        const acks = new Set(this.props.configuration.context.get('acknowledged-issue-numbers') ?? []);
        acks.add(Number(noticeId));
        this.props.configuration.context.set('acknowledged-issue-numbers', Array.from(acks));
        await this.props.configuration.saveContext();
    }
    async cliTelemetryStatus(args) {
        const canCollect = (0, collect_telemetry_1.canCollectTelemetry)(args, this.props.configuration.context);
        if (canCollect) {
            await this.ioHost.asIoHelper().defaults.info('CLI Telemetry is enabled. See https://docs.aws.amazon.com/cdk/v2/guide/cli-telemetry.html for ways to disable.');
        }
        else {
            await this.ioHost.asIoHelper().defaults.info('CLI Telemetry is disabled. See https://docs.aws.amazon.com/cdk/v2/guide/cli-telemetry.html for ways to enable.');
        }
    }
    async cliTelemetry(enable) {
        this.props.configuration.context.set('cli-telemetry', enable);
        await this.props.configuration.saveContext();
        await this.ioHost.asIoHelper().defaults.info(`Telemetry ${enable ? 'enabled' : 'disabled'}`);
    }
    async diff(options) {
        const stacks = await this.selectStacksForDiff(options.stackNames, options.exclusively);
        const strict = !!options.strict;
        const contextLines = options.contextLines || 3;
        const quiet = options.quiet || false;
        let diffs = 0;
        const parameterMap = buildParameterMap(options.parameters);
        if (options.templatePath !== undefined) {
            // Compare single stack against fixed template
            if (stacks.stackCount !== 1) {
                throw new toolkit_lib_1.ToolkitError('SingleStackRequired', 'Can only select one stack when comparing to fixed template. Use --exclusively to avoid selecting multiple stacks.');
            }
            if (!(await fs.pathExists(options.templatePath))) {
                throw new toolkit_lib_1.ToolkitError('TemplateNotFound', `There is no file at ${options.templatePath}`);
            }
            if (options.importExistingResources) {
                throw new toolkit_lib_1.ToolkitError('ImportWithTemplatePath', 'Can only use --import-existing-resources flag when comparing against deployed stacks.');
            }
            const template = (0, util_1.deserializeStructure)(await fs.readFile(options.templatePath, { encoding: 'utf-8' }));
            const formatter = new api_1.DiffFormatter({
                templateInfo: {
                    oldTemplate: template,
                    newTemplate: stacks.firstStack,
                },
            });
            if (options.securityOnly) {
                const securityDiff = formatter.formatSecurityDiff({ quiet });
                // Warn, count, and display the diff only if the reported changes are broadening permissions
                if (securityDiff.permissionChangeType === toolkit_lib_1.PermissionChangeType.BROADENING) {
                    await this.ioHost.asIoHelper().defaults.warn('This deployment will make potentially sensitive changes according to your current security approval level.\nPlease confirm you intend to make the following modifications:\n');
                    await this.ioHost.asIoHelper().defaults.info(securityDiff.formattedDiff);
                    diffs += securityDiff.numStacksWithChanges;
                }
            }
            else {
                const diff = formatter.formatStackDiff({
                    strict,
                    contextLines,
                    quiet,
                });
                diffs = diff.numStacksWithChanges;
                await this.ioHost.asIoHelper().defaults.info(diff.formattedDiff);
            }
        }
        else {
            const allMappings = options.includeMoves
                ? await (0, refactor_1.mappingsByEnvironment)(stacks.stackArtifacts, this.props.sdkProvider, true)
                : [];
            // Compare N stacks against deployed templates
            for (const stack of stacks.stackArtifacts) {
                const templateWithNestedStacks = await this.props.deployments.readCurrentTemplateWithNestedStacks(stack, options.compareAgainstProcessedTemplate);
                const currentTemplate = templateWithNestedStacks.deployedRootTemplate;
                const nestedStacks = templateWithNestedStacks.nestedStacks;
                const migrator = new api_1.ResourceMigrator({
                    deployments: this.props.deployments,
                    ioHelper: (0, api_private_1.asIoHelper)(this.ioHost, 'diff'),
                });
                const resourcesToImport = await migrator.tryGetResources(await this.props.deployments.resolveEnvironment(stack));
                if (resourcesToImport) {
                    (0, api_1.removeNonImportResources)(stack);
                }
                const changeSet = (options.method !== 'template')
                    ? await this.tryCreateDiffChangeSet(stack, options, parameterMap, resourcesToImport, quiet)
                    : undefined;
                const mappings = allMappings.find(m => m.environment.region === stack.environment.region && m.environment.account === stack.environment.account)?.mappings ?? {};
                const formatter = new api_1.DiffFormatter({
                    templateInfo: {
                        oldTemplate: currentTemplate,
                        newTemplate: stack,
                        changeSet,
                        isImport: !!resourcesToImport,
                        nestedStacks,
                        mappings,
                    },
                });
                if (options.securityOnly) {
                    const securityDiff = formatter.formatSecurityDiff({ quiet });
                    // Warn, count, and display the diff only if the reported changes are broadening permissions
                    if (securityDiff.permissionChangeType === toolkit_lib_1.PermissionChangeType.BROADENING) {
                        await this.ioHost.asIoHelper().defaults.warn('This deployment will make potentially sensitive changes according to your current security approval level.\nPlease confirm you intend to make the following modifications:\n');
                        await this.ioHost.asIoHelper().defaults.info(securityDiff.formattedDiff);
                        diffs += securityDiff.numStacksWithChanges;
                    }
                }
                else {
                    const diff = formatter.formatStackDiff({
                        strict,
                        contextLines,
                        quiet,
                    });
                    await this.ioHost.asIoHelper().defaults.info(diff.formattedDiff);
                    diffs += diff.numStacksWithChanges;
                }
            }
        }
        await this.ioHost.asIoHelper().defaults.info((0, node_util_1.format)('\n✨  Number of stacks with differences: %s\n', diffs));
        return diffs && options.fail ? 1 : 0;
    }
    /**
     * Try to create a diff changeset for the given stack.
     * Returns undefined if the stack cannot be accessed and changeSetOnly is not set.
     */
    async tryCreateDiffChangeSet(stack, options, parameterMap, resourcesToImport, quiet) {
        try {
            // we don't actually need to know if the stack exists here
            // we just use this to flush our any permissions issues and drop the result
            void await this.props.deployments.stackExists({
                stack,
                deployName: stack.stackName,
                tryLookupRole: true,
            });
        }
        catch (e) {
            if (options.method === 'change-set') {
                throw toolkit_lib_1.ToolkitError.withCause('DescribeStacksFailed', `Could not access stack '${stack.stackName}'. Please check your permissions or use '--method=auto' to allow falling back to a template diff.`, e);
            }
            await this.ioHost.asIoHelper().defaults.debug((0, util_1.formatErrorMessage)(e));
            if (!quiet) {
                await this.ioHost.asIoHelper().defaults.info(`Could not access stack '${stack.stackName}', falling back to template diff. Use '--method=change-set' to fail instead. Run with -v to see the reason.\n`);
            }
            return undefined;
        }
        return api_private_1.cfnApi.createDiffChangeSet((0, api_private_1.asIoHelper)(this.ioHost, 'diff'), {
            stack,
            uuid: (0, node_crypto_1.randomUUID)(),
            deployments: this.props.deployments,
            willExecute: false,
            sdkProvider: this.props.sdkProvider,
            parameters: Object.assign({}, parameterMap['*'], parameterMap[stack.stackName]),
            resourcesToImport,
            importExistingResources: options.importExistingResources,
            failOnError: options.method === 'change-set',
        });
    }
    async deploy(options) {
        if (options.watch) {
            return this.watch(options);
        }
        // set progress from options, this includes user and app config
        if (options.progress) {
            this.ioHost.stackProgress = options.progress;
        }
        // the ioHost uses this internally to determine if a confirmation
        // is actually needed, so it needs the same value we determine here.
        const requireApproval = options.requireApproval ?? cloud_assembly_schema_1.RequireApproval.BROADENING;
        this.ioHost.requireDeployApproval = requireApproval;
        // execute-change-set is a new flow that we can just delegate to toolkit-lib
        if (options.deploymentMethod?.method === 'execute-change-set') {
            await this.toolkit.deploy(this.props.cloudExecutable, {
                deploymentMethod: options.deploymentMethod,
                stacks: {
                    patterns: options.selector.patterns,
                    strategy: api_1.StackSelectionStrategy.PATTERN_MUST_MATCH_SINGLE,
                    expand: cloud_assembly_1.ExpandStackSelection.NONE,
                },
                roleArn: options.roleArn,
                forceDeployment: options.force,
                rollback: options.rollback,
                reuseAssets: options.reuseAssets,
                concurrency: options.concurrency,
                traceLogs: options.traceLogs,
                notificationArns: options.notificationArns,
                tags: options.tags,
                outputsFile: options.outputsFile,
                assetParallelism: options.assetParallelism,
                assetBuildConcurrency: options.assetBuildConcurrency,
                assetBuildTime: options.assetBuildTime,
                parameters: undefined, // parameters are only set during change set creation, so this is explicitly unset because change set already exists
            });
            return;
        }
        const startSynthTime = new Date().getTime();
        const stackCollection = await this.selectStacksForDeploy(options.selector, options.exclusively, options.cacheCloudAssembly, options.ignoreNoStacks);
        const elapsedSynthTime = new Date().getTime() - startSynthTime;
        await this.ioHost.asIoHelper().defaults.info(`\n✨  Synthesis time: ${(0, util_1.formatTime)(elapsedSynthTime)}s\n`);
        if (stackCollection.stackCount === 0) {
            await this.ioHost.asIoHelper().defaults.error('This app contains no stacks');
            return;
        }
        const migrator = new api_1.ResourceMigrator({
            deployments: this.props.deployments,
            ioHelper: (0, api_private_1.asIoHelper)(this.ioHost, 'deploy'),
        });
        await migrator.tryMigrateResources(stackCollection, {
            toolkitStackName: this.toolkitStackName,
            ...options,
        });
        if (options.deploymentMethod?.method === 'hotswap') {
            await this.ioHost.asIoHelper().defaults.warn('⚠️ The --hotswap and --hotswap-fallback flags deliberately introduce CloudFormation drift to speed up deployments');
            await this.ioHost.asIoHelper().defaults.warn('⚠️ They should only be used for development - never use them for your production Stacks!\n');
        }
        const stacks = stackCollection.stackArtifacts;
        const assetBuildTime = options.assetBuildTime ?? AssetBuildTime.ALL_BEFORE_DEPLOY;
        const prebuildAssets = assetBuildTime === AssetBuildTime.ALL_BEFORE_DEPLOY;
        const concurrency = options.concurrency || 1;
        if (concurrency > 1) {
            // always force "events" progress output when we have concurrency
            this.ioHost.stackProgress = deploy_1.StackActivityProgress.EVENTS;
            // ...but only warn if the user explicitly requested "bar" progress
            if (options.progress && options.progress != deploy_1.StackActivityProgress.EVENTS) {
                await this.ioHost.asIoHelper().defaults.warn('⚠️ The --concurrency flag only supports --progress "events". Switching to "events".');
            }
        }
        const stacksAndTheirAssetManifests = stacks.flatMap((stack) => [
            stack,
            ...stack.dependencies.filter(x => cxapi.AssetManifestArtifact.isAssetManifestArtifact(x)),
        ]);
        const workGraph = new api_1.WorkGraphBuilder((0, api_private_1.asIoHelper)(this.ioHost, 'deploy'), prebuildAssets).build(stacksAndTheirAssetManifests);
        // Unless we are running with '--force', skip already published assets
        if (!options.force) {
            await this.removePublishedAssets(workGraph, options);
        }
        const graphConcurrency = {
            'stack': concurrency,
            'asset-build': (options.assetParallelism ?? true) ? options.assetBuildConcurrency ?? 1 : 1, // This will be CPU-bound/memory bound, mostly matters for Docker builds
            'asset-publish': (options.assetParallelism ?? true) ? 8 : 1, // This will be I/O-bound, 8 in parallel seems reasonable
            'marker': 1,
        };
        const deploymentActions = new WorkGraphDeploymentActions(this.props.deployments, this.ioHost, this, {
            roleArn: options.roleArn,
            force: options.force,
            stackCount: stackCollection.stackCount,
            notificationArns: options.notificationArns,
            deploymentMethod: options.deploymentMethod,
            toolkitStackName: this.toolkitStackName,
            reuseAssets: options.reuseAssets,
            tags: options.tags,
            parameters: options.parameters,
            usePreviousParameters: options.usePreviousParameters,
            rollback: options.rollback,
            concurrency,
            requireApproval,
            assetParallelism: options.assetParallelism,
            extraUserAgent: options.extraUserAgent,
            cloudWatchLogMonitor: options.cloudWatchLogMonitor,
            sdkProvider: this.props.sdkProvider,
            express: options.express,
        });
        const startDeployTime = Date.now();
        await workGraph.doParallel(graphConcurrency, deploymentActions);
        if (options.outputsFile) {
            // If an outputs file has been specified, create the file path and write stack outputs to it once.
            // Outputs are written after all stacks have been deployed. If a stack deployment fails,
            // all of the outputs from successfully deployed stacks before the failure will still be written.
            await deploymentActions.writeOutputs(options.outputsFile);
        }
        // Add a timer on the COMMAND span for the full deployment wait time (not the same as the sum of all DEPLOY
        // spans because of parallelism).
        this.ioHost.telemetry?.commandSpan?.addTimer('totalDeployTime', Date.now() - startDeployTime);
        await this.ioHost.asIoHelper().defaults.info(`\n✨  Total time: ${(0, util_1.formatTime)(Date.now() - startSynthTime)}s\n`);
    }
    /**
     * Detect infrastructure drift for the given stack(s)
     */
    async drift(options) {
        const driftResults = await this.toolkit.drift(this.props.cloudExecutable, {
            stacks: {
                patterns: options.selector.patterns,
                strategy: options.selector.patterns.length > 0 ? api_1.StackSelectionStrategy.PATTERN_MATCH : api_1.StackSelectionStrategy.ALL_STACKS,
            },
        });
        const totalDrifts = Object.values(driftResults).reduce((total, current) => total + (current.numResourcesWithDrift ?? 0), 0);
        return totalDrifts > 0 && options.fail ? 1 : 0;
    }
    /**
     * Validate synthesized templates against policy rules
     */
    async validate(options) {
        // Implicitly switch 'debug' mode to true; more stack traces = more useful.
        this.props.cloudExecutable.switchOnDebugging();
        const result = await this.toolkit.validate(this.props.cloudExecutable, options);
        return result.conclusion === 'failure' ? 1 : 0;
    }
    /**
     * Diagnose errors
     */
    async diagnose(options) {
        // Implicitly switch 'debug' mode to true; more stack traces = more useful.
        this.props.cloudExecutable.switchOnDebugging();
        const results = await this.toolkit.diagnose(this.props.cloudExecutable, options);
        if (results.stacks.some(s => s.result.type !== 'no-problem')) {
            return 1;
        }
        return 0;
    }
    /**
     * Roll back the given stack or stacks.
     */
    async rollback(options) {
        const startSynthTime = new Date().getTime();
        const stackCollection = await this.selectStacksForDeploy(options.selector, true);
        const elapsedSynthTime = new Date().getTime() - startSynthTime;
        await this.ioHost.asIoHelper().defaults.info(`\n✨  Synthesis time: ${(0, util_1.formatTime)(elapsedSynthTime)}s\n`);
        if (stackCollection.stackCount === 0) {
            await this.ioHost.asIoHelper().defaults.error('No stacks selected');
            return;
        }
        let anyRollbackable = false;
        for (const stack of stackCollection.stackArtifacts) {
            await this.ioHost.asIoHelper().defaults.info('Rolling back %s', chalk_1.default.bold(stack.displayName));
            const startRollbackTime = new Date().getTime();
            try {
                const result = await this.props.deployments.rollbackStack({
                    stack,
                    roleArn: options.roleArn,
                    toolkitStackName: options.toolkitStackName,
                    orphanFailedResources: options.force,
                    validateBootstrapStackVersion: options.validateBootstrapStackVersion,
                    orphanLogicalIds: options.orphanLogicalIds,
                });
                if (!result.notInRollbackableState) {
                    anyRollbackable = true;
                }
                const elapsedRollbackTime = new Date().getTime() - startRollbackTime;
                await this.ioHost.asIoHelper().defaults.info(`\n✨  Rollback time: ${(0, util_1.formatTime)(elapsedRollbackTime).toString()}s\n`);
            }
            catch (e) {
                await this.ioHost.asIoHelper().defaults.error('\n ❌  %s failed: %s', chalk_1.default.bold(stack.displayName), (0, util_1.formatErrorMessage)(e));
                throw new toolkit_lib_1.ToolkitError('RollbackFailed', 'Rollback failed (use --force to orphan failing resources)');
            }
        }
        if (!anyRollbackable) {
            throw new toolkit_lib_1.ToolkitError('NoRollbackableStacks', 'No stacks were in a state that could be rolled back');
        }
    }
    async publishAssets(options) {
        await this.toolkit.publishAssets(this.props.cloudExecutable, options);
    }
    async watch(options) {
        const rootDir = path.dirname(path.resolve(user_configuration_1.PROJECT_CONFIG));
        const ioHelper = (0, api_private_1.asIoHelper)(this.ioHost, 'watch');
        await this.ioHost.asIoHelper().defaults.debug("root directory used for 'watch' is: %s", rootDir);
        const watchSettings = this.props.configuration.settings.get(['watch']);
        if (!watchSettings) {
            throw new toolkit_lib_1.ToolkitError('WatchConfigMissing', "Cannot use the 'watch' command without specifying at least one directory to monitor. " +
                'Make sure to add a "watch" key to your cdk.json');
        }
        // For the "include" subkey under the "watch" key, the behavior is:
        // 1. No "watch" setting? We error out.
        // 2. "watch" setting without an "include" key? We default to observing "./**".
        // 3. "watch" setting with an empty "include" key? We default to observing "./**".
        // 4. Non-empty "include" key? Just use the "include" key.
        // Note: We use '**' as the default pattern (not rootDir) because chokidar reports
        // file paths relative to cwd, and the ignored function uses picomatch which expects
        // glob patterns, not absolute paths.
        const watchIncludes = this.patternsArrayForWatch(watchSettings.include, {
            defaultPattern: '**',
            returnDefaultIfEmpty: true,
        });
        await this.ioHost.asIoHelper().defaults.debug("'include' patterns for 'watch': %s", watchIncludes);
        // For the "exclude" subkey under the "watch" key,
        // the behavior is to add some default excludes in addition to the ones specified by the user:
        // 1. The CDK output directory.
        // 2. Any file whose name starts with a dot.
        // 3. Any directory's content whose name starts with a dot.
        // 4. Any node_modules and its content (even if it's not a JS/TS project, you might be using a local aws-cli package)
        const outputDir = this.props.configuration.settings.get(['output']);
        const watchExcludes = this.patternsArrayForWatch(watchSettings.exclude, {
            defaultPattern: '',
            returnDefaultIfEmpty: false,
        }).concat(`${outputDir}/**`, '**/.*', '**/.*/**', '**/node_modules/**');
        await this.ioHost.asIoHelper().defaults.debug("'exclude' patterns for 'watch': %s", watchExcludes);
        // Since 'cdk deploy' is a relatively slow operation for a 'watch' process,
        // introduce a concurrency latch that tracks the state.
        // This way, if file change events arrive when a 'cdk deploy' is still executing,
        // we will batch them, and trigger another 'cdk deploy' after the current one finishes,
        // making sure 'cdk deploy's  always execute one at a time.
        // Here's a diagram showing the state transitions:
        // --------------                --------    file changed     --------------    file changed     --------------  file changed
        // |            |  ready event   |      | ------------------> |            | ------------------> |            | --------------|
        // | pre-ready  | -------------> | open |                     | deploying  |                     |   queued   |               |
        // |            |                |      | <------------------ |            | <------------------ |            | <-------------|
        // --------------                --------  'cdk deploy' done  --------------  'cdk deploy' done  --------------
        let latch = 'pre-ready';
        const cloudWatchLogMonitor = options.traceLogs ? new api_1.CloudWatchLogEventMonitor({
            ioHelper,
        }) : undefined;
        const deployAndWatch = async () => {
            latch = 'deploying';
            await cloudWatchLogMonitor?.deactivate();
            await this.invokeDeployFromWatch(options, cloudWatchLogMonitor);
            // If latch is still 'deploying' after the 'await', that's fine,
            // but if it's 'queued', that means we need to deploy again
            while (latch === 'queued') {
                // TypeScript doesn't realize latch can change between 'awaits',
                // and thinks the above 'while' condition is always 'false' without the cast
                latch = 'deploying';
                await this.ioHost.asIoHelper().defaults.info("Detected file changes during deployment. Invoking 'cdk deploy' again");
                await this.invokeDeployFromWatch(options, cloudWatchLogMonitor);
            }
            latch = 'open';
            await cloudWatchLogMonitor?.activate();
        };
        // Create ignore matcher for chokidar v4 compatibility
        // Chokidar v4 removed glob pattern support, so we use picomatch to filter files
        // We pass rootDir because chokidar v4 passes absolute paths to the ignored callback
        const shouldIgnore = (0, api_private_1.createIgnoreMatcher)({
            include: watchIncludes,
            exclude: watchExcludes,
            rootDir,
        });
        chokidar
            .watch('.', {
            ignored: shouldIgnore,
            cwd: rootDir,
        })
            .on('ready', async () => {
            latch = 'open';
            await this.ioHost.asIoHelper().defaults.debug("'watch' received the 'ready' event. From now on, all file changes will trigger a deployment");
            await this.ioHost.asIoHelper().defaults.info("Triggering initial 'cdk deploy'");
            await deployAndWatch();
        })
            .on('all', async (event, filePath) => {
            if (!isFileEvent(event)) {
                return; // Ignore non-file events like 'error', 'raw', 'ready', 'all'
            }
            if (latch === 'pre-ready') {
                await this.ioHost.asIoHelper().defaults.info(`'watch' is observing ${event === 'addDir' ? 'directory' : 'the file'} '%s' for changes`, filePath);
            }
            else if (latch === 'open') {
                await this.ioHost.asIoHelper().defaults.info("Detected change to '%s' (type: %s). Triggering 'cdk deploy'", filePath, event);
                await deployAndWatch();
            }
            else {
                // this means latch is either 'deploying' or 'queued'
                latch = 'queued';
                await this.ioHost.asIoHelper().defaults.info("Detected change to '%s' (type: %s) while 'cdk deploy' is still running. " +
                    'Will queue for another deployment after this one finishes', filePath, event);
            }
        });
    }
    async orphan(options) {
        await this.toolkit.orphan(this.props.cloudExecutable, {
            constructPaths: options.constructPath,
            roleArn: options.roleArn,
            toolkitStackName: options.toolkitStackName,
        });
    }
    async import(options) {
        const stacks = await this.selectStacksForDeploy(options.selector, true, true, false);
        // set progress from options, this includes user and app config
        if (options.progress) {
            this.ioHost.stackProgress = options.progress;
        }
        if (stacks.stackCount > 1) {
            throw new toolkit_lib_1.ToolkitError('AmbiguousStackSelection', `Stack selection is ambiguous, please choose a specific stack for import [${stacks.stackArtifacts.map((x) => x.id).join(', ')}]`);
        }
        if (!process.stdout.isTTY && !options.resourceMappingFile && !options.resourceMappingInline) {
            throw new toolkit_lib_1.ToolkitError('ResourceMappingRequired', '--resource-mapping or --resource-mapping-inline is required when input is not a terminal');
        }
        const stack = stacks.stackArtifacts[0];
        await this.ioHost.asIoHelper().defaults.info(chalk_1.default.bold(stack.displayName));
        const resourceImporter = new api_1.ResourceImporter(stack, {
            deployments: this.props.deployments,
            ioHelper: (0, api_private_1.asIoHelper)(this.ioHost, 'import'),
        });
        const { additions, hasNonAdditions, diffFormatter } = await resourceImporter.discoverImportableResources(options.force);
        // If there are non-addition changes (e.g. after orphan, hardcoded refs differ from Fn::GetAtt),
        // warn the user and ask for confirmation unless --force was given.
        if (hasNonAdditions && !options.force) {
            const ioHelper = this.ioHost.asIoHelper();
            await ioHelper.defaults.info(`The following resources have pending updates that will be reconciled with a ${chalk_1.default.blueBright('cdk deploy')} after import:`);
            const { formattedDiff } = diffFormatter.formatStackDiff();
            await ioHelper.defaults.info(formattedDiff);
            const confirmed = await ioHelper.requestResponse(api_private_1.IO.CDK_TOOLKIT_I7010.req('Perform import?', { motivation: 'Confirm import with pending drift' }));
            if (!confirmed) {
                throw new toolkit_lib_1.AbortError('ImportAborted', 'Import cancelled');
            }
        }
        if (additions.length === 0) {
            await this.ioHost.asIoHelper().defaults.warn('%s: no new resources compared to the currently deployed stack, skipping import.', chalk_1.default.bold(stack.displayName));
            return;
        }
        // Prepare a mapping of physical resources to CDK constructs
        let actualImport;
        if (options.resourceMappingInline) {
            actualImport = await resourceImporter.loadResourceIdentifiers(additions, options.resourceMappingInline);
        }
        else if (options.resourceMappingFile) {
            actualImport = await resourceImporter.loadResourceIdentifiersFromFile(additions, options.resourceMappingFile);
        }
        else {
            actualImport = await resourceImporter.askForResourceIdentifiers(additions);
        }
        if (actualImport.importResources.length === 0) {
            await this.ioHost.asIoHelper().defaults.warn('No resources selected for import.');
            return;
        }
        // If "--create-resource-mapping" option was passed, write the resource mapping to the given file and exit
        if (options.recordResourceMapping) {
            const outputFile = options.recordResourceMapping;
            fs.ensureFileSync(outputFile);
            await fs.writeJson(outputFile, actualImport.resourceMap, {
                spaces: 2,
                encoding: 'utf8',
            });
            await this.ioHost.asIoHelper().defaults.info('%s: mapping file written.', outputFile);
            return;
        }
        // Import the resources according to the given mapping
        await this.ioHost.asIoHelper().defaults.info('%s: importing resources into stack...', chalk_1.default.bold(stack.displayName));
        const tags = (0, api_private_1.tagsForStack)(stack);
        await resourceImporter.importResourcesFromMap(actualImport, {
            roleArn: options.roleArn,
            tags,
            deploymentMethod: options.deploymentMethod,
            usePreviousParameters: true,
            rollback: options.rollback,
        });
        // Notify user of next steps
        if (actualImport.importResources.length < additions.length) {
            await this.ioHost.asIoHelper().defaults.warn(`Some resources were skipped. Run another ${chalk_1.default.blueBright('cdk import')} or a ${chalk_1.default.blueBright('cdk deploy')} to bring the stack up-to-date with your CDK app definition.`);
        }
        else if (hasNonAdditions) {
            // After orphan→import, the deployed template still has hardcoded values that differ from
            // the synth'd template's Fn::GetAtt/Ref intrinsics. A deploy updates the template to match the CDK app.
            if (options.force) {
                await this.ioHost.asIoHelper().defaults.info(`Import complete. Run ${chalk_1.default.blueBright('cdk deploy')} to update the stack to match your CDK app.`);
            }
            else {
                const deployNow = await this.ioHost.asIoHelper().requestResponse(api_private_1.IO.CDK_TOOLKIT_I7010.req(`Finish with a ${chalk_1.default.blueBright('cdk deploy')} now?`, { motivation: 'Update stack to match CDK app after import' }));
                if (deployNow) {
                    await this.deploy({
                        selector: options.selector,
                        toolkitStackName: options.toolkitStackName,
                        roleArn: options.roleArn,
                        deploymentMethod: options.deploymentMethod,
                    });
                }
                else {
                    await this.ioHost.asIoHelper().defaults.info(`Import complete. Remember to run ${chalk_1.default.blueBright('cdk deploy')} to update the stack to match your CDK app.`);
                }
            }
        }
        else {
            await this.ioHost.asIoHelper().defaults.info(`Import operation complete. We recommend you run a ${chalk_1.default.blueBright('drift detection')} operation ` +
                'to confirm your CDK app resource definitions are up-to-date. Read more here: ' +
                chalk_1.default.underline.blueBright('https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/detect-drift-stack.html'));
        }
    }
    async destroy(options) {
        // Keep the "deployed" wording when a destroy runs as part of a deploy.
        const action = options.fromDeploy ? 'deploy' : 'destroy';
        if ((options.concurrency || 1) > 1) {
            this.ioHost.stackProgress = deploy_1.StackActivityProgress.EVENTS;
        }
        // The destroy action runs through toolkit-lib.
        // Use listeners to keep the messages as they were before.
        this.ioHost.on(api_private_1.IO.CDK_TOOLKIT_I1001, () => ({ preventDefault: true })); // Starting Synthesis (trace)
        this.ioHost.on(api_private_1.IO.CDK_TOOLKIT_I1000, () => ({ preventDefault: true })); // ✨ Synthesis time (info)
        this.ioHost.on(api_private_1.IO.CDK_TOOLKIT_I7101, () => ({ preventDefault: true })); // Starting Destroy (trace)
        this.ioHost.on(api_private_1.IO.CDK_TOOLKIT_I7001, () => ({ preventDefault: true })); // per-stack Destroy time (trace)
        this.ioHost.on(api_private_1.IO.CDK_TOOLKIT_I7000, () => ({ preventDefault: true })); // ✨ Destroy time (info)
        // The success line was `info` in the historical `cdk destroy`, not `result`.
        this.ioHost.on(api_private_1.IO.CDK_TOOLKIT_I7900, () => ({ level: 'info' })); // ✅ <stack>: destroyed
        // toolkit-lib logs a declined confirmation (E7010) and returns gracefully.
        // The CLI surfaces a decline as a non-zero, soft exit instead: throwing from
        // the listener both suppresses the log and aborts the command (the top-level
        // renders `AbortError` as "Deletion cancelled").
        this.ioHost.on(api_private_1.IO.CDK_TOOLKIT_E7010, () => {
            throw new toolkit_lib_1.AbortError('DestroyAborted', 'Deletion cancelled');
        });
        if (options.force) {
            this.ioHost.respondOnce(api_private_1.IO.CDK_TOOLKIT_I7010, true);
        }
        try {
            // eslint-disable-next-line @typescript-eslint/ban-ts-comment
            // @ts-ignore - `_destroyWithAction` is private; the CLI sets the action label.
            await this.toolkit._destroyWithAction(this.props.cloudExecutable, action, {
                stacks: {
                    patterns: options.selector.patterns,
                    strategy: options.selector.allTopLevel
                        ? api_1.StackSelectionStrategy.MAIN_ASSEMBLY
                        : options.selector.patterns.length > 0
                            ? api_1.StackSelectionStrategy.PATTERN_MATCH
                            : api_1.StackSelectionStrategy.ONLY_SINGLE,
                    expand: options.exclusively ? cloud_assembly_1.ExpandStackSelection.NONE : cloud_assembly_1.ExpandStackSelection.DOWNSTREAM,
                },
                roleArn: options.roleArn,
                concurrency: options.concurrency,
                express: options.express,
            });
        }
        finally {
            this.ioHost.removeAllListeners();
        }
    }
    async list(selectors, options = {}) {
        this.ioHost.rewriteOnce(api_private_1.IO.CDK_TOOLKIT_I2901, (msg) => (0, list_stacks_1.formatStackList)(msg.data.stacks, options));
        // cdk ls stdout is the stack listing only. Suppress the info status lines synthesis emits next
        // to it: the synth-time line (I1000) and the dependency-expansion note (I1002).
        this.ioHost.once(api_private_1.IO.CDK_TOOLKIT_I1000, () => ({ preventDefault: true }));
        this.ioHost.once(api_private_1.IO.CDK_TOOLKIT_I1002, () => ({ preventDefault: true }));
        await this.toolkit.list(this.props.cloudExecutable, {
            stacks: selectors.length > 0
                ? { patterns: selectors, strategy: api_1.StackSelectionStrategy.PATTERN_MATCH, expand: cloud_assembly_1.ExpandStackSelection.UPSTREAM }
                : undefined,
        });
        return 0; // exit-code
    }
    /**
     * Synthesize the given set of stacks (called when the user runs 'cdk synth')
     *
     * INPUT: Stack names can be supplied using a glob filter. If no stacks are
     * given, all stacks from the application are implicitly selected.
     *
     * OUTPUT: If more than one stack ends up being selected, an output directory
     * should be supplied, where the templates will be written.
     */
    async synth(stackNames, exclusively, quiet, autoValidate, json) {
        const stacks = await this.selectStacksForDiff(stackNames, exclusively, autoValidate);
        // if we have a single stack, print it to STDOUT
        if (stacks.stackCount === 1) {
            if (!quiet) {
                await printSerializedObject(this.ioHost.asIoHelper(), (0, util_1.obscureTemplate)(stacks.firstStack.template), json ?? false);
            }
            // In CI mode, non-error messages go to stdout. When we just printed the
            // template to stdout, skip the flags message to preserve the contract that
            // `cdk synth` output is valid YAML. When quiet (no template printed) or
            // non-CI (flags go to stderr), it's safe to show.
            if (quiet || !this.ioHost.isCI) {
                await displayFlagsMessage(this.ioHost.asIoHelper(), this.toolkit, this.props.cloudExecutable);
            }
            return undefined;
        }
        // not outputting template to stdout, let's explain things to the user a little bit...
        await this.ioHost.asIoHelper().defaults.info(chalk_1.default.green(`Successfully synthesized to ${chalk_1.default.blue(path.resolve(stacks.assembly.directory))}`));
        await this.ioHost.asIoHelper().defaults.info(`Supply a stack id (${stacks.stackArtifacts.map((s) => chalk_1.default.green(s.hierarchicalId)).join(', ')}) to display its template.`);
        await displayFlagsMessage(this.ioHost.asIoHelper(), this.toolkit, this.props.cloudExecutable);
        return undefined;
    }
    /**
     * Bootstrap the CDK Toolkit stack in the accounts used by the specified stack(s).
     *
     * @param userEnvironmentSpecs - environment names that need to have toolkit support
     *             provisioned, as a glob filter. If none is provided, all stacks are implicitly selected.
     * @param options - The name, role ARN, bootstrapping parameters, etc. to be used for the CDK Toolkit stack.
     */
    async bootstrap(userEnvironmentSpecs, options) {
        const bootstrapper = new bootstrap_1.Bootstrapper(options.source, (0, api_private_1.asIoHelper)(this.ioHost, 'bootstrap'));
        // If there is an '--app' argument and an environment looks like a glob, we
        // select the environments from the app. Otherwise, use what the user said.
        const environments = await this.defineEnvironments(userEnvironmentSpecs);
        const limit = pLimit(20);
        // eslint-disable-next-line @cdklabs/promiseall-no-unbounded-parallelism
        await Promise.all(environments.map((environment) => limit(async () => {
            await this.ioHost.asIoHelper().defaults.info(chalk_1.default.green(' ⏳  Bootstrapping environment %s...'), chalk_1.default.blue(environment.name));
            try {
                const result = await bootstrapper.bootstrapEnvironment(environment, this.props.sdkProvider, options);
                const message = result.noOp
                    ? '✅  Environment %s bootstrapped (no changes).'
                    : '✅  Environment %s bootstrapped.';
                await this.ioHost.asIoHelper().defaults.info(chalk_1.default.green(message), chalk_1.default.blue(environment.name));
                if (options.express) {
                    const warning = (0, api_private_1.formatExpressStabilizationWarning)(result.stabilizingResources, 'bootstrap');
                    if (warning) {
                        await this.ioHost.asIoHelper().notify(api_private_1.IO.CDK_TOOLKIT_W9902.msg(warning));
                    }
                }
            }
            catch (e) {
                await this.ioHost.asIoHelper().defaults.error(' ❌  Environment %s failed bootstrapping: %s', chalk_1.default.blue(environment.name), e);
                throw e;
            }
        })));
    }
    /**
     * Garbage collects assets from a CDK app's environment
     * @param options - Options for Garbage Collection
     */
    async garbageCollect(userEnvironmentSpecs, options) {
        const environments = await this.defineEnvironments(userEnvironmentSpecs);
        for (const environment of environments) {
            await this.ioHost.asIoHelper().defaults.info(chalk_1.default.green(' ⏳  Garbage Collecting environment %s...'), chalk_1.default.blue(environment.name));
            const gc = new api_1.GarbageCollector({
                sdkProvider: this.props.sdkProvider,
                ioHelper: (0, api_private_1.asIoHelper)(this.ioHost, 'gc'),
                resolvedEnvironment: environment,
                bootstrapStackName: options.bootstrapStackName,
                rollbackBufferDays: options.rollbackBufferDays,
                createdBufferDays: options.createdBufferDays,
                action: options.action ?? 'full',
                type: options.type ?? 'all',
                confirm: options.confirm ?? true,
            });
            await gc.garbageCollect();
        }
    }
    async defineEnvironments(userEnvironmentSpecs) {
        // By default, glob for everything
        const environmentSpecs = userEnvironmentSpecs.length > 0 ? [...userEnvironmentSpecs] : ['**'];
        // Partition into globs and non-globs (this will mutate environmentSpecs).
        const globSpecs = (0, util_1.partition)(environmentSpecs, cxapp_1.looksLikeGlob);
        if (globSpecs.length > 0 && !this.props.cloudExecutable.hasApp) {
            if (userEnvironmentSpecs.length > 0) {
                // User did request this glob
                throw new toolkit_lib_1.ToolkitError('InvalidEnvironmentGlob', `'${globSpecs}' is not an environment name. Specify an environment name like 'aws://123456789012/us-east-1', or run in a directory with 'cdk.json' to use wildcards.`);
            }
            else {
                // User did not request anything
                throw new toolkit_lib_1.ToolkitError('EnvironmentRequired', "Specify an environment name like 'aws://123456789012/us-east-1', or run in a directory with 'cdk.json'.");
            }
        }
        const environments = [...(0, cxapp_1.environmentsFromDescriptors)(environmentSpecs)];
        // If there is an '--app' argument, select the environments from the app.
        if (this.props.cloudExecutable.hasApp) {
            environments.push(...(await (0, cxapp_1.globEnvironmentsFromStacks)(await this.selectStacksForList([]), globSpecs, this.props.sdkProvider)));
        }
        return environments;
    }
    /**
     * Migrates a CloudFormation stack/template to a CDK app
     * @param options - Options for CDK app creation
     */
    async migrate(options) {
        await this.ioHost.asIoHelper().defaults.warn('This command is an experimental feature.');
        const language = options.language?.toLowerCase() ?? 'typescript';
        const environment = (0, migrate_1.setEnvironment)(options.account, options.region);
        let generateTemplateOutput;
        let cfn;
        let templateToDelete;
        try {
            // if neither fromPath nor fromStack is provided, generate a template using cloudformation
            const scanType = (0, migrate_1.parseSourceOptions)(options.fromPath, options.fromStack, options.stackName).source;
            if (scanType == migrate_1.TemplateSourceOptions.SCAN) {
                generateTemplateOutput = await (0, migrate_1.generateTemplate)({
                    stackName: options.stackName,
                    filters: options.filter,
                    fromScan: options.fromScan,
                    sdkProvider: this.props.sdkProvider,
                    environment: environment,
                    ioHelper: this.ioHost.asIoHelper(),
                });
                templateToDelete = generateTemplateOutput.templateId;
            }
            else if (scanType == migrate_1.TemplateSourceOptions.PATH) {
                const templateBody = (0, migrate_1.readFromPath)(options.fromPath);
                const parsedTemplate = (0, util_1.deserializeStructure)(templateBody);
                const templateId = parsedTemplate.Metadata?.AWSToolsMetrics?.IaC_Generator?.toString();
                if (templateId) {
                    // if we have a template id, we can call describe generated template to get the resource identifiers
                    // resource metadata, and template source to generate the template
                    cfn = new migrate_1.CfnTemplateGeneratorProvider(await (0, migrate_1.buildCfnClient)(this.props.sdkProvider, environment), this.ioHost.asIoHelper());
                    const generatedTemplateSummary = await cfn.describeGeneratedTemplate(templateId);
                    generateTemplateOutput = (0, migrate_1.buildGeneratedTemplateOutput)(generatedTemplateSummary, templateBody, generatedTemplateSummary.GeneratedTemplateId);
                }
                else {
                    generateTemplateOutput = {
                        migrateJson: {
                            templateBody: templateBody,
                            source: 'localfile',
                        },
                    };
                }
            }
            else if (scanType == migrate_1.TemplateSourceOptions.STACK) {
                const template = await (0, migrate_1.readFromStack)(options.stackName, this.props.sdkProvider, environment);
                if (!template) {
                    throw new toolkit_lib_1.ToolkitError('StackTemplateNotFound', `No template found for stack-name: ${options.stackName}`);
                }
                generateTemplateOutput = {
                    migrateJson: {
                        templateBody: template,
                        source: options.stackName,
                    },
                };
            }
            else {
                // We shouldn't ever get here, but just in case.
                throw new toolkit_lib_1.ToolkitError('InvalidSourceOption', `Invalid source option provided: ${scanType}`);
            }
            const stack = (0, migrate_1.generateStack)(generateTemplateOutput.migrateJson.templateBody, options.stackName, language);
            await this.ioHost.asIoHelper().defaults.info(chalk_1.default.green(' ⏳  Generating CDK app for %s...'), chalk_1.default.blue(options.stackName));
            await (0, migrate_1.generateCdkApp)(this.ioHost.asIoHelper(), options.stackName, stack, language, options.outputPath, options.compress);
            if (generateTemplateOutput) {
                (0, migrate_1.writeMigrateJsonFile)(options.outputPath, options.stackName, generateTemplateOutput.migrateJson);
            }
            if ((0, migrate_1.isThereAWarning)(generateTemplateOutput)) {
                await this.ioHost.asIoHelper().defaults.warn(' ⚠️  Some resources could not be migrated completely. Please review the README.md file for more information.');
                (0, migrate_1.appendWarningsToReadme)(`${path.join(options.outputPath ?? process.cwd(), options.stackName)}/README.md`, generateTemplateOutput.resources);
            }
        }
        catch (e) {
            await this.ioHost.asIoHelper().defaults.error(' ❌  Migrate failed for `%s`: %s', options.stackName, e.message);
            throw e;
        }
        finally {
            if (templateToDelete) {
                if (!cfn) {
                    cfn = new migrate_1.CfnTemplateGeneratorProvider(await (0, migrate_1.buildCfnClient)(this.props.sdkProvider, environment), this.ioHost.asIoHelper());
                }
                if (!process.env.MIGRATE_INTEG_TEST) {
                    await cfn.deleteGeneratedTemplate(templateToDelete);
                }
            }
        }
    }
    async refactor(options) {
        if (options.revert && !options.overrideFile) {
            throw new toolkit_lib_1.ToolkitError('RevertRequiresOverrideFile', 'The --revert option can only be used with the --override-file option.');
        }
        try {
            const patterns = options.stacks?.patterns ?? [];
            await this.toolkit.refactor(this.props.cloudExecutable, {
                dryRun: options.dryRun,
                stacks: {
                    patterns: patterns,
                    strategy: patterns.length > 0 ? api_1.StackSelectionStrategy.PATTERN_MATCH : api_1.StackSelectionStrategy.ALL_STACKS,
                },
                force: options.force,
                additionalStackNames: options.additionalStackNames,
                overrides: readOverrides(options.overrideFile, options.revert),
                roleArn: options.roleArn,
            });
        }
        catch (e) {
            await this.ioHost.asIoHelper().defaults.error(e.message);
            throw e;
        }
        return 0;
        function readOverrides(filePath, revert = false) {
            if (filePath == null) {
                return [];
            }
            if (!fs.pathExistsSync(filePath)) {
                throw new toolkit_lib_1.ToolkitError('MappingFileNotFound', `The mapping file ${filePath} does not exist`);
            }
            const groups = (0, refactor_1.parseMappingGroups)(fs.readFileSync(filePath).toString('utf-8'));
            return revert
                ? groups.map((group) => ({
                    ...group,
                    resources: Object.fromEntries(Object.entries(group.resources ?? {}).map(([src, dst]) => [dst, src])),
                }))
                : groups;
        }
    }
    async selectStacksForList(patterns) {
        const assembly = await this.assembly();
        const stacks = await assembly.selectStacks({ patterns }, { defaultBehavior: cxapp_1.DefaultSelection.AllStacks });
        // No validation
        return stacks;
    }
    async selectStacksForDeploy(selector, exclusively, cacheCloudAssembly, ignoreNoStacks) {
        const assembly = await this.assembly(cacheCloudAssembly);
        const stacks = await assembly.selectStacks(selector, {
            extend: exclusively ? cloud_assembly_1.ExtendedStackSelection.None : cloud_assembly_1.ExtendedStackSelection.Upstream,
            defaultBehavior: cxapp_1.DefaultSelection.OnlySingle,
            ignoreNoStacks,
        });
        this.validateStacksSelected(stacks, selector.patterns);
        await this.validateStacks(assembly, stacks);
        return stacks;
    }
    async selectStacksForDiff(stackNames, exclusively, autoValidate) {
        const assembly = await this.assembly();
        const selectedForDiff = await assembly.selectStacks({ patterns: stackNames }, {
            extend: exclusively ? cloud_assembly_1.ExtendedStackSelection.None : cloud_assembly_1.ExtendedStackSelection.Upstream,
            defaultBehavior: cxapp_1.DefaultSelection.MainAssembly,
        });
        const allStacks = await this.selectStacksForList([]);
        const autoValidateStacks = autoValidate
            ? allStacks.filter((art) => art.validateOnSynth ?? false)
            : new cloud_assembly_1.StackCollection(assembly, []);
        this.validateStacksSelected(selectedForDiff.concat(autoValidateStacks), stackNames);
        await this.validateStacks(assembly, selectedForDiff.concat(autoValidateStacks));
        return selectedForDiff;
    }
    /**
     * Validate the stacks for errors and warnings according to the CLI's current settings
     */
    async validateStacks(assembly, stacks) {
        const failAt = this.validateMetadataFailAt();
        await (0, api_private_1.throwIfValidationFailures)(assembly, stacks, failAt, this.ioHost.asIoHelper());
    }
    validateMetadataFailAt() {
        let failAt = 'error';
        if (this.props.ignoreErrors) {
            failAt = 'none';
        }
        if (this.props.strict) {
            failAt = 'warn';
        }
        return failAt;
    }
    /**
     * Validate that if a user specified a stack name there exists at least 1 stack selected
     */
    validateStacksSelected(stacks, stackNames) {
        if (stackNames.length != 0 && stacks.stackCount == 0) {
            throw new toolkit_lib_1.ToolkitError('NoStacksMatched', `No stacks match the name(s) ${stackNames}`);
        }
    }
    /**
     * Select a single stack by its name
     */
    async selectSingleStackByName(stackName) {
        const assembly = await this.assembly();
        const stacks = await assembly.selectStacks({ patterns: [stackName] }, {
            extend: cloud_assembly_1.ExtendedStackSelection.None,
            defaultBehavior: cxapp_1.DefaultSelection.None,
        });
        // Could have been a glob so check that we evaluated to exactly one
        if (stacks.stackCount > 1) {
            throw new toolkit_lib_1.ToolkitError('MultipleStacksMatched', `This command requires exactly one stack and we matched more than one: ${stacks.stackIds}`);
        }
        return assembly.stackById(stacks.firstStack.id);
    }
    assembly(cacheCloudAssembly) {
        return this.props.cloudExecutable.synthesize(cacheCloudAssembly);
    }
    patternsArrayForWatch(patterns, options) {
        const patternsArray = patterns !== undefined ? (Array.isArray(patterns) ? patterns : [patterns]) : [];
        return patternsArray.length > 0 ? patternsArray : options.returnDefaultIfEmpty ? [options.defaultPattern] : [];
    }
    async invokeDeployFromWatch(options, cloudWatchLogMonitor) {
        const deployOptions = {
            ...options,
            requireApproval: cloud_assembly_schema_1.RequireApproval.NEVER,
            // if 'watch' is called by invoking 'cdk deploy --watch',
            // we need to make sure to not call 'deploy' with 'watch' again,
            // as that would lead to a cycle
            watch: false,
            cloudWatchLogMonitor,
            cacheCloudAssembly: false,
            extraUserAgent: `cdk-watch/hotswap-${options.deploymentMethod?.method === 'hotswap' ? 'on' : 'off'}`,
            concurrency: options.concurrency,
        };
        try {
            await this.deploy(deployOptions);
        }
        catch {
            // just continue - deploy will show the error
        }
    }
    /**
     * Remove the asset publishing and building from the work graph for assets that are already in place
     */
    async removePublishedAssets(graph, options) {
        await graph.removeUnnecessaryAssets(assetNode => this.props.deployments.isSingleAssetPublished(assetNode.assetManifest, assetNode.asset, {
            stack: assetNode.parentStack,
            roleArn: options.roleArn,
            stackName: assetNode.parentStack.stackName,
        }));
    }
}
exports.CdkToolkit = CdkToolkit;
/**
 * Print a serialized object (YAML or JSON) to stdout.
 */
async function printSerializedObject(ioHelper, obj, json) {
    await ioHelper.defaults.result((0, util_1.serializeStructure)(obj, json));
}
function buildParameterMap(parameters) {
    const parameterMap = { '*': {} };
    for (const key in parameters) {
        if (parameters.hasOwnProperty(key)) {
            const [stack, parameter] = key.split(':', 2);
            if (!parameter) {
                parameterMap['*'][stack] = parameters[key];
            }
            else {
                if (!parameterMap[stack]) {
                    parameterMap[stack] = {};
                }
                parameterMap[stack][parameter] = parameters[key];
            }
        }
    }
    return parameterMap;
}
/**
 * Ask the user for a yes/no confirmation
 *
 * Automatically fail the confirmation in case we're in a situation where the confirmation
 * cannot be interactively obtained from a human at the keyboard.
 */
async function askUserConfirmation(ioHost, req, abortErrorCode, abortMessage) {
    await ioHost.withCorkedLogging(async () => {
        // The IoHost returns the answer rather than aborting; abort here on decline.
        const confirmed = await ioHost.asIoHelper().requestResponse(req);
        if (!confirmed) {
            throw new toolkit_lib_1.AbortError(abortErrorCode, abortMessage);
        }
    });
}
/**
 * Display a warning if there are flags that are different from the recommended value
 *
 * This happens if both of the following are true:
 *
 * - The user didn't configure the value
 * - The default value for the flag (unconfiguredBehavesLike) is different from the recommended value
 */
async function displayFlagsMessage(ioHost, toolkit, cloudExecutable) {
    const flags = await toolkit.flags(cloudExecutable);
    // The "unconfiguredBehavesLike" information got added later. If none of the flags have this information,
    // we don't have enough information to reliably display this information without scaring users, so don't do anything.
    if (flags.every(flag => flag.unconfiguredBehavesLike === undefined)) {
        return;
    }
    const unconfiguredFlags = operations_1.FlagOperations.filterNeedsAttention(flags);
    const numUnconfigured = unconfiguredFlags.length;
    if (numUnconfigured > 0) {
        await ioHost.defaults.warn(`${numUnconfigured} feature flags are not configured. Run 'cdk flags --unstable=flags' to learn more.`);
    }
}
/**
 * This class implements the callbacks for the parallel work graph executions.
 *
 * All this code used to be inlined into CdkToolkit#deploy(), but it was getting
 * too much to live in there.
 */
class WorkGraphDeploymentActions {
    deployments;
    ioHost;
    stackOperations;
    options;
    stackOutputs = {};
    assetSpans = {};
    stackCounter = 1;
    constructor(deployments, ioHost, stackOperations, options) {
        this.deployments = deployments;
        this.ioHost = ioHost;
        this.stackOperations = stackOperations;
        this.options = options;
    }
    async marker(markerNode) {
        // For assets, emit IoHost messages that have been defined to start and end asset spans, which will lead to events in telemetry.
        switch (markerNode.marker.type) {
            case 'start-asset': {
                const assetSpan = await this.obtainSpan(markerNode.marker.asset);
                // Count the asset type, both on the asset-specific span and on the global COMMAND span.
                if (markerNode.marker.asset.type === 'file') {
                    this.ioHost.telemetry?.commandSpan?.incCounter('fileAsset');
                    assetSpan.incCounter('fileAsset');
                }
                else if (markerNode.marker.asset.type === 'docker') {
                    this.ioHost.telemetry?.commandSpan?.incCounter('dockerAsset');
                    assetSpan.incCounter('dockerAsset');
                }
                break;
            }
            case 'end-asset': {
                const assetSpan = await this.obtainSpan(markerNode.marker.asset);
                await assetSpan.end({});
                break;
            }
        }
    }
    async obtainSpan(assetEntry) {
        const span = this.assetSpans[assetEntry.id.assetId];
        if (span) {
            return span;
        }
        const assetSpan = await this.ioHost.asIoHelper().span(messages_1.CLI_PRIVATE_SPAN.ASSET).begin({});
        this.assetSpans[assetEntry.id.assetId] = assetSpan;
        return assetSpan;
    }
    async buildAsset(assetNode) {
        const env_1 = { stack: [], error: void 0, hasError: false };
        try {
            // Measure asset build time. Both on the global COMMAND span as well as on the asset-specific span.
            // (parallelism may cause SUM(buildAssetST) >= totalDeployTime) on the COMMAND span, indicate with [S]ub[T]ask suffix).
            const _invokeTimer = __addDisposableResource(env_1, this.ioHost.telemetry?.commandSpan?.startTimer('buildAssetST'), false);
            const _assetTimer = __addDisposableResource(env_1, (await this.obtainSpan(assetNode.asset)).startTimer('buildAsset'), false);
            await this.deployments.buildSingleAsset(assetNode.assetManifestArtifact, assetNode.assetManifest, assetNode.asset, {
                stack: assetNode.parentStack,
                roleArn: this.options.roleArn,
                stackName: assetNode.parentStack.stackName,
            });
        }
        catch (e_1) {
            env_1.error = e_1;
            env_1.hasError = true;
        }
        finally {
            __disposeResources(env_1);
        }
    }
    async publishAsset(assetNode) {
        const env_2 = { stack: [], error: void 0, hasError: false };
        try {
            // Measure asset publish time. Both on the global COMMAND span as well as on the asset-specific span.
            // (parallelism may cause SUM(publishAssetST) >= totalDeployTime) on the COMMAND span, indicate with [S]ub[T]ask suffix).
            const _invokeTimer = __addDisposableResource(env_2, this.ioHost.telemetry?.commandSpan?.startTimer('publishAssetST'), false);
            const _assetTimer = __addDisposableResource(env_2, (await this.obtainSpan(assetNode.asset)).startTimer('publishAsset'), false);
            await this.deployments.publishSingleAsset(assetNode.assetManifest, assetNode.asset, {
                stack: assetNode.parentStack,
                roleArn: this.options.roleArn,
                stackName: assetNode.parentStack.stackName,
                forcePublish: this.options.force,
            });
        }
        catch (e_2) {
            env_2.error = e_2;
            env_2.hasError = true;
        }
        finally {
            __disposeResources(env_2);
        }
    }
    async deployStack(stackNode) {
        // Counter for totalDeployedStacks, to match 'totalDeployTime' timer on the COMMAND span.
        this.ioHost.telemetry?.commandSpan?.incCounter('totalDeployedStacks');
        const stack = stackNode.stack;
        if (this.options.stackCount !== 1) {
            await this.ioHost.asIoHelper().defaults.info(chalk_1.default.bold(stack.displayName));
        }
        if (!stack.environment) {
            // eslint-disable-next-line @stylistic/max-len
            throw new toolkit_lib_1.ToolkitError('MissingEnvironment', `Stack ${stack.displayName} does not define an environment, and AWS credentials could not be obtained from standard locations or no region was configured.`);
        }
        const resourceCount = Object.keys(stack.template.Resources || {}).length;
        if (resourceCount === 0) {
            // The generated stack has no resources
            if (!(await this.deployments.stackExists({ stack }))) {
                await this.ioHost.asIoHelper().defaults.warn('%s: stack has no resources, skipping deployment.', chalk_1.default.bold(stack.displayName));
            }
            else {
                await this.ioHost.asIoHelper().defaults.warn('%s: stack has no resources, deleting existing stack.', chalk_1.default.bold(stack.displayName));
                await this.stackOperations.destroy({
                    selector: { patterns: [stack.hierarchicalId] },
                    exclusively: true,
                    force: true,
                    roleArn: this.options.roleArn,
                    fromDeploy: true,
                });
            }
            return;
        }
        // Following are the same semantics we apply with respect to Notification ARNs (dictated by the SDK)
        //
        //  - undefined  =>  cdk ignores it, as if it wasn't supported (allows external management).
        //  - []:        =>  cdk manages it, and the user wants to wipe it out.
        //  - ['arn-1']  =>  cdk manages it, and the user wants to set it to ['arn-1'].
        const notificationArns = (!!this.options.notificationArns || !!stack.notificationArns)
            ? (this.options.notificationArns ?? []).concat(stack.notificationArns ?? [])
            : undefined;
        for (const notificationArn of notificationArns ?? []) {
            if (!(0, util_1.validateSnsTopicArn)(notificationArn)) {
                throw new toolkit_lib_1.ToolkitError('InvalidSnsTopicArn', `Notification arn ${notificationArn} is not a valid arn for an SNS topic`);
            }
        }
        const parameterMap = buildParameterMap(this.options.parameters);
        // Deploy options that are shared between change set creation and execution
        const sharedDeployOptions = {
            stack,
            deployName: stack.stackName,
            roleArn: this.options.roleArn,
            toolkitStackName: this.options.toolkitStackName,
            reuseAssets: this.options.reuseAssets,
            tags: (this.options.tags?.length ? this.options.tags : (0, api_private_1.tagsForStack)(stack)),
            forceDeployment: this.options.force,
            parameters: Object.assign({}, parameterMap['*'], parameterMap[stack.stackName]),
            usePreviousParameters: this.options.usePreviousParameters,
            rollback: this.options.rollback,
            notificationArns,
            extraUserAgent: this.options.extraUserAgent,
            assetParallelism: this.options.assetParallelism,
            express: this.options.express,
        };
        // When using change-set method, always create the change set upfront.
        // This gives us an accurate diff for approval and avoids creating it twice.
        // For non-executing deployments (prepare-change-set), this is the final result.
        const prepareResult = (0, deploy_private_1.isChangeSetDeployment)(this.options.deploymentMethod)
            ? await this.deployments.prepareStack({
                ...sharedDeployOptions,
                deploymentMethod: this.options.deploymentMethod,
                cleanupOnNoOp: (0, deploy_private_1.isExecutingChangeSetDeployment)(this.options.deploymentMethod),
            })
            : undefined;
        // Also skip the approval flow when the prepared change set is a no-op —
        // there is nothing for the user to approve. Outputs, stack ARN, and
        // timings are still emitted via the normal no-op deploy path below.
        if (this.options.requireApproval !== cloud_assembly_schema_1.RequireApproval.NEVER && !prepareResult?.noOp) {
            const currentTemplate = await this.deployments.readCurrentTemplate(stack);
            const formatter = new api_1.DiffFormatter({
                templateInfo: {
                    oldTemplate: currentTemplate,
                    newTemplate: stack,
                    changeSet: prepareResult?.changeSet,
                },
            });
            const securityDiff = formatter.formatSecurityDiff();
            if (requiresApproval(this.options.requireApproval, securityDiff.permissionChangeType)) {
                const hasSecurityChanges = securityDiff.permissionChangeType !== toolkit_lib_1.PermissionChangeType.NONE;
                const motivation = hasSecurityChanges
                    ? '"--require-approval" is enabled and stack includes security-sensitive updates'
                    : `"--require-approval" is set to '${cloud_assembly_schema_1.RequireApproval.ANYCHANGE}'`;
                const diffOutput = hasSecurityChanges ? securityDiff.formattedDiff : formatter.formatStackDiff().formattedDiff;
                await this.ioHost.asIoHelper().defaults.info(diffOutput);
                try {
                    await askUserConfirmation(this.ioHost, api_private_1.IO.CDK_TOOLKIT_I5060.req(`${motivation}: Do you wish to deploy these changes?`, {
                        motivation,
                        concurrency: this.options.concurrency,
                        permissionChangeType: securityDiff.permissionChangeType,
                        templateDiffs: formatter.diffs,
                    }), 'DeployAborted', 'Deployment cancelled');
                }
                catch (e) {
                    if (prepareResult?.changeSet?.ChangeSetName) {
                        await this.deployments.cleanupChangeSet(stack, prepareResult.changeSet.ChangeSetName);
                    }
                    throw e;
                }
            }
        }
        const stackIndex = this.stackCounter++;
        await this.ioHost.asIoHelper().defaults.info(`${chalk_1.default.bold(stack.displayName)}: deploying... [${stackIndex}/${this.options.stackCount}]`);
        const startDeployTime = new Date().getTime();
        // There is already a startDeployTime constant, but that does not work with telemetry.
        // We should integrate the two in the future
        const deploySpan = await this.ioHost.asIoHelper().span(messages_1.CLI_PRIVATE_SPAN.DEPLOY).begin({});
        deploySpan.incCounter('resources', resourceCount);
        this.ioHost.telemetry?.commandSpan?.incCounter('totalDeployedResources', resourceCount);
        let error;
        let elapsedDeployTime = 0;
        try {
            // The prepare result is final if the change set was empty (noOp) or
            // the deployment method is non-executing (prepare-change-set).
            const prepareIsFinal = prepareResult && (prepareResult.noOp || (0, deploy_private_1.isNonExecutingChangeSetDeployment)(this.options.deploymentMethod));
            let deployResult = prepareIsFinal ? prepareResult : undefined;
            // Start with user config for rollback,
            // but it might change if we encounter a failed state.
            let rollback = this.options.rollback;
            // We limit the loop to 2 iterations max as defensive programming.
            // Should not be possible to happen.
            let iteration = 0;
            while (!deployResult) {
                if (++iteration > 2) {
                    throw new toolkit_lib_1.ToolkitError('DeployLoopUnstable', 'This loop should have stabilized in 2 iterations, but didn\'t. If you are seeing this error, please report it at https://github.com/aws/aws-cdk/issues/new/choose');
                }
                const r = await this.deployments.deployStack({
                    ...sharedDeployOptions,
                    // On the first iteration, execute the prepared change set.
                    // On retries (after rollback), create a new change set since the old one is gone.
                    deploymentMethod: iteration === 1 && (0, deploy_private_1.isExecutingChangeSetDeployment)(this.options.deploymentMethod)
                        ? (0, deploy_private_1.toExecuteChangeSetDeployment)(this.options.deploymentMethod)
                        : this.options.deploymentMethod,
                    rollback,
                });
                switch (r.type) {
                    case 'did-deploy-stack':
                        deployResult = r;
                        break;
                    case 'failpaused-need-rollback-first': {
                        const motivation = r.reason === 'replacement'
                            ? `Stack is in a paused fail state (${r.status}) and change includes a replacement which cannot be deployed with "--no-rollback"`
                            : `Stack is in a paused fail state (${r.status}) and command line arguments do not include "--no-rollback"`;
                        if (this.options.force) {
                            await this.ioHost.asIoHelper().defaults.warn(`${motivation}. Rolling back first (--force).`);
                        }
                        else {
                            await askUserConfirmation(this.ioHost, api_private_1.IO.CDK_TOOLKIT_I5050.req(`${motivation}. Roll back first and then proceed with deployment`, {
                                motivation,
                                concurrency: this.options.concurrency,
                            }), 'RollbackAborted', 'Rollback cancelled');
                        }
                        // Perform a rollback
                        await this.stackOperations.rollback({
                            selector: { patterns: [stack.hierarchicalId] },
                            toolkitStackName: this.options.toolkitStackName,
                            force: this.options.force,
                        });
                        // Go around through the 'while' loop again but switch rollback to true.
                        rollback = true;
                        break;
                    }
                    case 'replacement-requires-rollback': {
                        const motivation = 'Change includes a replacement which cannot be deployed with "--no-rollback"';
                        if (this.options.force) {
                            await this.ioHost.asIoHelper().defaults.warn(`${motivation}. Proceeding with deployment with rollback enabled (--force).`);
                        }
                        else {
                            await askUserConfirmation(this.ioHost, api_private_1.IO.CDK_TOOLKIT_I5050.req(`${motivation}. Perform a deployment with rollback enabled`, {
                                concurrency: this.options.concurrency,
                                motivation,
                            }), 'ReplacementRollbackAborted', 'Rollback cancelled');
                        }
                        // Go around through the 'while' loop again but switch rollback to true.
                        rollback = true;
                        break;
                    }
                    default:
                        throw new toolkit_lib_1.ToolkitError('UnexpectedDeployResult', `Unexpected result type from deployStack: ${JSON.stringify(r)}. If you are seeing this error, please report it at https://github.com/aws/aws-cdk/issues/new/choose`);
                }
            }
            const message = deployResult.noOp
                ? '✅  %s (no changes)'
                : '✅  %s';
            await this.ioHost.asIoHelper().defaults.info(chalk_1.default.green('\n' + message), stack.displayName);
            elapsedDeployTime = new Date().getTime() - startDeployTime;
            await this.ioHost.asIoHelper().defaults.info(`\n✨  Deployment time: ${(0, util_1.formatTime)(elapsedDeployTime)}s\n`);
            if (this.options.express) {
                const warning = (0, api_private_1.formatExpressStabilizationWarning)(deployResult.stabilizingResources, 'deploy');
                if (warning) {
                    await this.ioHost.asIoHelper().notify(api_private_1.IO.CDK_TOOLKIT_W5902.msg(warning));
                }
            }
            if (Object.keys(deployResult.outputs).length > 0) {
                await this.ioHost.asIoHelper().defaults.info('Outputs:');
                this.stackOutputs[stack.stackName] = deployResult.outputs;
            }
            for (const name of Object.keys(deployResult.outputs).sort()) {
                const value = deployResult.outputs[name];
                await this.ioHost.asIoHelper().defaults.info(`${chalk_1.default.cyan(stack.id)}.${chalk_1.default.cyan(name)} = ${chalk_1.default.underline(chalk_1.default.cyan(value))}`);
            }
            await this.ioHost.asIoHelper().defaults.info('Stack ARN:');
            await this.ioHost.asIoHelper().defaults.result(deployResult.stackArn);
        }
        catch (e) {
            // It has to be exactly this string because an integration test tests for
            // "bold(stackname) failed: ResourceNotReady: <error>"
            const code = toolkit_lib_1.ToolkitError.isToolkitError(e) ? e.name : 'DeployStackFailed'; // Formerly 'DeployFailed'
            const newMessage = [`❌  ${chalk_1.default.bold(stack.stackName)} failed:`, ...(e.name ? [`${e.name}:`] : []), e.message].join(' ');
            const wrappedError = new toolkit_lib_1.ToolkitError(code, newMessage);
            error = {
                name: (0, error_1.cdkCliErrorName)(wrappedError),
            };
            throw wrappedError;
        }
        finally {
            await deploySpan.end({ error });
            if (this.options.cloudWatchLogMonitor) {
                const foundLogGroupsResult = await (0, api_1.findCloudWatchLogGroups)(this.options.sdkProvider, (0, api_private_1.asIoHelper)(this.ioHost, 'deploy'), stack);
                this.options.cloudWatchLogMonitor.addLogGroups(foundLogGroupsResult.env, foundLogGroupsResult.sdk, foundLogGroupsResult.logGroupNames);
            }
        }
    }
    async writeOutputs(outputsFile) {
        fs.ensureFileSync(outputsFile);
        await fs.writeJson(outputsFile, this.stackOutputs, {
            spaces: 2,
            encoding: 'utf8',
        });
    }
}
/**
 * Determine if manual approval is required or not. Requires approval for
 * - RequireApproval.ANYCHANGE
 * - RequireApproval.BROADENING and the changes are indeed broadening permissions
 */
function requiresApproval(requireApproval, permissionChangeType) {
    return requireApproval === cloud_assembly_schema_1.RequireApproval.ANYCHANGE ||
        requireApproval === cloud_assembly_schema_1.RequireApproval.BROADENING && permissionChangeType === toolkit_lib_1.PermissionChangeType.BROADENING;
}
