// aws-cdk/lib/commands/doctor.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.doctor = doctor;
const process = __importStar(require("process"));
const cxapi = __importStar(require("@aws-cdk/cx-api"));
const chalk_1 = __importDefault(require("chalk"));
const display_version_1 = require("../cli/display-version");
const version_1 = require("../cli/version");
async function doctor({ ioHelper, settings }) {
    let exitStatus = 0;
    for (const verification of verifications) {
        if (!await verification(ioHelper, settings)) {
            exitStatus = -1;
        }
    }
    await (0, display_version_1.displayVersionMessage)(ioHelper);
    return exitStatus;
}
const verifications = [
    displayVersionInformation,
    displayCliConfiguration,
    displayAwsEnvironmentVariables,
    displayCdkEnvironmentVariables,
];
// ### Verifications ###
async function displayCliConfiguration(ioHelper, settings) {
    const verbosity = Number(settings?.get(['verbose']) ?? 0);
    const verbose = verbosity ? `extra verbosity (-${'v'.repeat(verbosity)})` : 'normal verbosity';
    const debug = [];
    if (Boolean(settings?.get(['debugApp']))) {
        debug.push('CDK app');
    }
    if (Boolean(settings?.get(['debugCli']))) {
        debug.push('CLI');
    }
    const debugging = debug.length ? `debugging ${debug.join(' & ')}` : 'no debugging';
    await ioHelper.defaults.info(chalk_1.default.gray.italic(`${verbose}, ${debugging}`));
    return true;
}
async function displayVersionInformation(ioHelper) {
    await ioHelper.defaults.info(`ℹ️ CDK CLI Version: ${chalk_1.default.green((0, version_1.versionWithBuild)())}`);
    return true;
}
async function displayAwsEnvironmentVariables(ioHelper) {
    const keys = Object.keys(process.env).filter(s => s.startsWith('AWS_'));
    if (keys.length === 0) {
        await ioHelper.defaults.info('ℹ️ No AWS environment variables');
        return true;
    }
    await ioHelper.defaults.info('ℹ️ AWS environment variables:');
    for (const key of keys) {
        await ioHelper.defaults.info(`  - ${chalk_1.default.blue(key)} = ${chalk_1.default.green(anonymizeAwsVariable(key, process.env[key]))}`);
    }
    return true;
}
async function displayCdkEnvironmentVariables(ioHelper) {
    const keys = Object.keys(process.env).filter(s => s.startsWith('CDK_'));
    if (keys.length === 0) {
        await ioHelper.defaults.info('ℹ️ No CDK environment variables');
        return true;
    }
    await ioHelper.defaults.info('ℹ️ CDK environment variables:');
    let healthy = true;
    for (const key of keys.sort()) {
        if (key === cxapi.CONTEXT_ENV || key === cxapi.CONTEXT_OVERFLOW_LOCATION_ENV || key === cxapi.OUTDIR_ENV) {
            await ioHelper.defaults.info(`  - ${chalk_1.default.red(key)} = ${chalk_1.default.green(process.env[key])} (⚠️ reserved for use by the CDK toolkit)`);
            healthy = false;
        }
        else {
            await ioHelper.defaults.info(`  - ${chalk_1.default.blue(key)} = ${chalk_1.default.green(process.env[key])}`);
        }
    }
    return healthy;
}
function anonymizeAwsVariable(name, value) {
    if (name === 'AWS_ACCESS_KEY_ID') {
        return value.slice(0, 4) + '<redacted>';
    } // Show ASIA/AKIA key type, but hide identifier
    if (name === 'AWS_SECRET_ACCESS_KEY' || name === 'AWS_SESSION_TOKEN' || name === 'AWS_SECURITY_TOKEN') {
        return '<redacted>';
    }
    return value;
}
