// aws-cdk/lib/cxapp/cloud-executable.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.CloudExecutable = void 0;
const toolkit_lib_1 = require("@aws-cdk/toolkit-lib");
const cloud_assembly_1 = require("./cloud-assembly");
const api_private_1 = require("../../lib/api-private");
const singleton_plugin_host_1 = require("../cli/singleton-plugin-host");
const error_1 = require("../cli/telemetry/error");
const messages_1 = require("../cli/telemetry/messages");
const contextproviders = __importStar(require("../context-providers"));
/**
 * Represent the Cloud Executable and the synthesis we can do on it
 */
class CloudExecutable {
    props;
    _cloudAssembly;
    constructor(props) {
        this.props = props;
    }
    async produce() {
        const synthesisResult = await this.synthesize(true);
        // We must return an `IReadableCloudAssembly` here, but this Cloud Assembly is only used in the context
        // of the CLI and `cli.ts` currently manages its own locking in the "synthesizer" callback function.
        //
        // All the lock-related functions are therefore no-ops.
        return new api_private_1.BorrowedAssembly(synthesisResult.assembly);
    }
    /**
     * Return whether there is an app command from the configuration
     */
    get hasApp() {
        return !!this.props.configuration.settings.get(['app']);
    }
    /**
     * Switch on debugging for the cloud executable.
     *
     * This will cause it to log more stack traces and other information that will make error
     * reports more useful (at the cost of increased execution time).
     */
    switchOnDebugging() {
        this.props.configuration.settings.temporarilyMutable((settings) => {
            settings.set(['debug'], true);
        });
    }
    /**
     * Synthesize a set of stacks.
     *
     * @param cacheCloudAssembly - whether to cache the Cloud Assembly after it has been first synthesized.
     *   This is 'true' by default, and only set to 'false' for 'cdk watch',
     *   which needs to re-synthesize the Assembly each time it detects a change to the project files
     */
    async synthesize(cacheCloudAssembly = true) {
        if (!this._cloudAssembly || !cacheCloudAssembly) {
            this._cloudAssembly = await this.doSynthesize();
        }
        return this._cloudAssembly;
    }
    async doSynthesize() {
        // We may need to run the cloud executable multiple times in order to satisfy all missing context
        // (When the executable runs, it will tell us about context it wants to use
        // but it missing. We'll then look up the context and run the executable again, and
        // again, until it doesn't complain anymore or we've stopped making progress).
        let previouslyMissingKeys;
        const synthSpan = await this.props.ioHelper.span(messages_1.CLI_PRIVATE_SPAN.SYNTH_ASSEMBLY).begin({});
        let error;
        try {
            while (true) {
                const assembly = await this.props.synthesizer(this.props.sdkProvider, this.props.configuration);
                if (assembly.manifest.missing && assembly.manifest.missing.length > 0) {
                    const missingKeys = missingContextKeys(assembly.manifest.missing);
                    if (!this.canLookup) {
                        throw new toolkit_lib_1.ToolkitError('ContextLookupsDisabled', 'Context lookups have been disabled. '
                            + 'Make sure all necessary context is already in \'cdk.context.json\' by running \'cdk synth\' on a machine with sufficient AWS credentials and committing the result. '
                            + `Missing context keys: '${Array.from(missingKeys).join(', ')}'`);
                    }
                    let tryLookup = true;
                    if (previouslyMissingKeys && setsEqual(missingKeys, previouslyMissingKeys)) {
                        await this.props.ioHelper.defaults.debug('Not making progress trying to resolve environmental context. Giving up.');
                        tryLookup = false;
                    }
                    previouslyMissingKeys = missingKeys;
                    if (tryLookup) {
                        const lookupsTimer = synthSpan.startTimer('lookups');
                        try {
                            await this.props.ioHelper.defaults.debug('Some context information is missing. Fetching...');
                            const updates = await contextproviders.provideContextValues(assembly.manifest.missing, this.props.sdkProvider, singleton_plugin_host_1.GLOBAL_PLUGIN_HOST, this.props.ioHelper);
                            for (const [key, value] of Object.entries(updates)) {
                                this.props.configuration.context.set(key, value);
                            }
                            // Cache the new context to disk
                            await this.props.configuration.saveContext();
                        }
                        finally {
                            lookupsTimer.stop();
                        }
                        // Execute again
                        continue;
                    }
                }
                (0, api_private_1.countAssemblyResults)(synthSpan, assembly);
                return new cloud_assembly_1.CloudAssembly(assembly, this.props.ioHelper);
            }
        }
        catch (e) {
            error = {
                name: (0, error_1.cdkCliErrorName)(e),
            };
            throw e;
        }
        finally {
            await synthSpan.end({ error });
        }
    }
    get canLookup() {
        return !!(this.props.configuration.settings.get(['lookups']) ?? true);
    }
}
exports.CloudExecutable = CloudExecutable;
/**
 * Return all keys of missing context items
 */
function missingContextKeys(missing) {
    return new Set((missing || []).map(m => m.key));
}
function setsEqual(a, b) {
    if (a.size !== b.size) {
        return false;
    }
    for (const x of a) {
        if (!b.has(x)) {
            return false;
        }
    }
    return true;
}
