// aws-cdk/lib/cli/telemetry/sink/funnel.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Funnel = void 0;
const toolkit_lib_1 = require("@aws-cdk/toolkit-lib");
/**
 * A funnel is a combination of one or more sinks.
 * The sink functions are executed in parallel, and a maximum of 5
 * sinks are supported per funnel.
 */
class Funnel {
    sinks;
    constructor(props) {
        if (props.sinks.length > 5) {
            throw new toolkit_lib_1.ToolkitError('MaxSinksExceeded', `Funnel class supports a maximum of 5 parallel sinks, got ${props.sinks.length} sinks.`);
        }
        this.sinks = props.sinks;
    }
    async emit(event) {
        // Funnel class enforces a maximum of 5 parallel sinks
        // eslint-disable-next-line @cdklabs/promiseall-no-unbounded-parallelism
        await Promise.all(this.sinks.map(sink => sink.emit(event)));
    }
    async flush() {
        // Funnel class enforces a maximum of 5 parallel sinks
        // eslint-disable-next-line @cdklabs/promiseall-no-unbounded-parallelism
        await Promise.all(this.sinks.map(sink => sink.flush()));
    }
}
exports.Funnel = Funnel;
