// aws-cdk/lib/cli/util/yargs-helpers.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.isCI = void 0;
exports.yargsNegativeAlias = yargsNegativeAlias;
exports.yargsImplies = yargsImplies;
exports.cliVersion = cliVersion;
exports.browserForPlatform = browserForPlatform;
const version_1 = require("../version");
var ci_1 = require("../util/ci");
Object.defineProperty(exports, "isCI", { enumerable: true, get: function () { return ci_1.isCI; } });
/**
 * yargs middleware to negate an option if a negative alias is provided
 * E.g. `-R` will imply `--rollback=false`
 *
 * @param optionToNegate - The name of the option to negate, e.g. `rollback`
 * @param negativeAlias - The alias that should negate the option, e.g. `R`
 * @returns a middleware function that can be passed to yargs
 */
function yargsNegativeAlias(negativeAlias, optionToNegate) {
    return (argv) => {
        // if R in argv && argv[R]
        // then argv[rollback] = false
        if (negativeAlias in argv && argv[negativeAlias]) {
            argv[optionToNegate] = false;
        }
        return argv;
    };
}
/**
 * yargs middleware to switch on other boolean options when a flag is set.
 * E.g. `--debug` implies `--debug-app` and `--debug-cli`.
 *
 * @param flag - The camelCase name of the flag that implies the others, e.g. `debug`
 * @param impliedOptions - The camelCase names of the options to switch on, e.g. `['debugApp', 'debugCli']`
 * @returns a middleware function that can be passed to yargs
 */
function yargsImplies(flag, impliedOptions) {
    return (argv) => {
        if (argv[flag]) {
            for (const implied of impliedOptions) {
                argv[implied] = true;
            }
        }
        return argv;
    };
}
/**
 * Returns the current version of the CLI
 * @returns the current version of the CLI
 */
function cliVersion() {
    return (0, version_1.versionWithBuild)();
}
/**
 * Returns the default browser command for the current platform
 * @returns the default browser command for the current platform
 */
function browserForPlatform() {
    switch (process.platform) {
        case 'darwin':
            return 'open %u';
        case 'win32':
            return 'start %u';
        default:
            return 'xdg-open %u';
    }
}
