// aws-cdk/lib/cli/telemetry/messages.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.CLI_PRIVATE_SPAN = exports.CLI_PRIVATE_IO = void 0;
const make = __importStar(require("../../api-private"));
/**
 * Private message types specific to the CLI
 */
exports.CLI_PRIVATE_IO = {
    CDK_CLI_I1000: make.trace({
        code: 'CDK_CLI_I1000',
        description: 'Cloud Execution is starting',
        interface: 'EventStart',
    }),
    CDK_CLI_I1001: make.trace({
        code: 'CDK_CLI_I1001',
        description: 'Cloud Executable Result',
        interface: 'EventResult',
    }),
    CDK_CLI_I2000: make.trace({
        code: 'CDK_CLI_I2000',
        description: 'Command has started',
        interface: 'EventStart',
    }),
    CDK_CLI_I2001: make.trace({
        code: 'CDK_CLI_I2001',
        description: 'Command has finished executing',
        interface: 'EventResult',
    }),
    CDK_CLI_I3000: make.trace({
        code: 'CDK_CLI_I3000',
        description: 'Deploy has started',
        interface: 'EventStart',
    }),
    CDK_CLI_I3001: make.trace({
        code: 'CDK_CLI_I3001',
        description: 'Deploy has finished',
        interface: 'EventResult',
    }),
    CDK_CLI_I3002: make.trace({
        code: 'CDK_CLI_I3002',
        description: 'Start asset building and publishing',
        interface: 'EventStart',
    }),
    CDK_CLI_I3003: make.trace({
        code: 'CDK_CLI_I3003',
        description: 'Finished asset building and publishing',
        interface: 'EventResult',
    }),
};
/**
 * Payload type of the end message must extend Duration
 */
exports.CLI_PRIVATE_SPAN = {
    SYNTH_ASSEMBLY: {
        name: 'Synthesis',
        start: exports.CLI_PRIVATE_IO.CDK_CLI_I1000,
        end: exports.CLI_PRIVATE_IO.CDK_CLI_I1001,
    },
    COMMAND: {
        name: 'Command',
        start: exports.CLI_PRIVATE_IO.CDK_CLI_I2000,
        end: exports.CLI_PRIVATE_IO.CDK_CLI_I2001,
    },
    DEPLOY: {
        name: 'Deploy',
        start: exports.CLI_PRIVATE_IO.CDK_CLI_I3000,
        end: exports.CLI_PRIVATE_IO.CDK_CLI_I3001,
    },
    ASSET: {
        name: 'Asset',
        start: exports.CLI_PRIVATE_IO.CDK_CLI_I3002,
        end: exports.CLI_PRIVATE_IO.CDK_CLI_I3003,
    },
};
