// aws-cdk/lib/cli/telemetry/session.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TelemetrySession = void 0;
exports.isValidWrapperUserAgent = isValidWrapperUserAgent;
const node_crypto_1 = require("node:crypto");
const toolkit_lib_1 = require("@aws-cdk/toolkit-lib");
const installation_id_1 = require("./installation-id");
const library_version_1 = require("./library-version");
const sanitation_1 = require("./sanitation");
const ci_systems_1 = require("../ci-systems");
const messages_1 = require("../telemetry/messages");
const ci_1 = require("../util/ci");
const version_1 = require("../version");
const error_1 = require("./error");
const telemetry_state_1 = require("./telemetry-state");
const ABORTED_ERROR_MESSAGE = '__CDK-Toolkit__Aborted';
/**
 * Valid user agent prefixes that are allowed to report through CDK_CLI_USERAGENT.
 * This creates a mechanism to report user agents we control.
 */
const VALID_USER_AGENTS = ['aws-blocks'];
class TelemetrySession {
    props;
    ioHost;
    client;
    _sessionInfo;
    _commandSpan;
    _nextEventCounters;
    count = 0;
    loadTime;
    constructor(props) {
        this.props = props;
        this.ioHost = props.ioHost;
        this.client = props.client;
    }
    /**
     * The span that represents the CLI invocation.
     *
     * In the code, this span is named COMMAND but the matching event type
     * in telemetry will be INVOKE.
     *
     * Will be emitted exactly once, at the end of the CLI operation.
     */
    get commandSpan() {
        return this._commandSpan;
    }
    async begin() {
        // sanitize the raw cli input
        const { path, parameters } = (0, sanitation_1.sanitizeCommandLineArguments)(this.props.arguments);
        this._sessionInfo = {
            identifiers: {
                installationId: await (0, installation_id_1.getOrCreateInstallationId)(this.ioHost.asIoHelper()),
                sessionId: (0, node_crypto_1.randomUUID)(),
                telemetryVersion: '2.0',
                cdkCliVersion: (0, version_1.versionNumber)(),
                cdkLibraryVersion: await (0, library_version_1.getLibraryVersion)(this.ioHost.asIoHelper()),
            },
            event: {
                command: {
                    path,
                    parameters,
                    config: {
                        context: (0, sanitation_1.sanitizeContext)(this.props.context),
                        ...(isValidWrapperUserAgent(process.env.CDK_CLI_USERAGENT)
                            ? { cdkCliUserAgent: { [process.env.CDK_CLI_USERAGENT]: true } }
                            : {}),
                    },
                },
            },
            environment: {
                ci: (0, ci_1.isCI)() || Boolean((0, ci_systems_1.detectCiSystem)()),
                os: {
                    platform: process.platform,
                    release: process.release.name,
                },
                nodeVersion: process.version,
            },
            project: {},
        };
        // If SIGINT has a listener installed, its default behavior will be removed (Node.js will no longer exit).
        // This ensures that on SIGINT we process safely close the telemetry session before exiting.
        process.on('SIGINT', async () => {
            try {
                await this.end({
                    name: error_1.USER_INTERRUPTED_CODE,
                    message: ABORTED_ERROR_MESSAGE,
                });
            }
            catch (e) {
                await this.ioHost.defaults.trace(`Ending Telemetry failed: ${e.message}`);
            }
            process.exit(1);
        });
        // Begin the session span
        this._commandSpan = await this.ioHost.asIoHelper().span(messages_1.CLI_PRIVATE_SPAN.COMMAND).begin({});
    }
    async attachRegion(region) {
        this.sessionInfo.identifiers = {
            ...this.sessionInfo.identifiers,
            region,
        };
    }
    /**
     * Attach a language guess
     */
    attachLanguage(language) {
        // Don't want to crash accidentally
        if (!this._sessionInfo) {
            return;
        }
        if (language) {
            mutable(this.sessionInfo.project).language = language;
        }
    }
    /**
     * Attach our best guess at running under an agent or not
     */
    attachAgent(isAgent) {
        // Don't want to crash accidentally
        if (!this._sessionInfo) {
            return;
        }
        mutable(this.sessionInfo.environment).agent = isAgent;
    }
    /**
     * Temporarily attach counters for the next event operation.
     *
     * They may be committed to the sent telemetry later.
     */
    attachCountersToNextEvent(counters) {
        this._nextEventCounters = counters;
    }
    /**
     * Set the load time (will be emitted with the COMMAND span)
     */
    attachLoadTime(loadTime) {
        this.loadTime = loadTime;
        this._commandSpan?.addTimer('load', loadTime);
    }
    /**
     * Mark when the actual CLI operation starts
     *
     * Emitted as part of the COMMAND span.
     */
    markOperationStart() {
        if (this.loadTime) {
            this._commandSpan?.addTimer('init', performance.now() - this.loadTime);
        }
    }
    /**
     * Attach the CDK library version
     *
     * By default the telemetry will guess at the CDK library version if it so
     * happens that the CDK project is an NPM project and the CDK CLI is executed
     * in the root of NPM project with `aws-cdk-lib` available in `node_modules`.
     * This may succeed or may fail.
     *
     * Once we have produced and loaded the cloud assembly more accurate
     * information becomes available that we can add in.
     */
    attachCdkLibVersion(libVersion) {
        // Don't want to crash accidentally
        if (!this._sessionInfo) {
            return;
        }
        mutable(this.sessionInfo.identifiers).cdkLibraryVersion = libVersion;
    }
    /**
     * When the command is complete, so is the CliIoHost. Ends the span of the entire CliIoHost
     * and notifies with an optional error message in the data.
     */
    async end(error) {
        await this._commandSpan?.end({ error });
        // Ideally span.end() should no-op if called twice, but that is not the case right now
        this._commandSpan = undefined;
        await this.client.flush();
    }
    async emit(event) {
        this.count += 1;
        const counters = {
            ...this._nextEventCounters,
            ...event.counters,
        };
        this._nextEventCounters = undefined;
        if (event.eventType == 'DEPLOY') {
            await this.trackDeployStatistics(event.error === undefined, counters);
        }
        return this.client.emit({
            event: {
                command: this.sessionInfo.event.command,
                state: getState(event.error),
                eventType: event.eventType,
            },
            identifiers: {
                ...this.sessionInfo.identifiers,
                eventId: `${this.sessionInfo.identifiers.sessionId}:${this.count}`,
                timestamp: new Date().toISOString(),
            },
            environment: this.sessionInfo.environment,
            project: this.sessionInfo.project,
            duration: {
                total: event.duration,
            },
            ...(event.error ? {
                error: {
                    name: event.error.name,
                },
            } : {}),
            ...(Object.keys(counters).length > 0 ? { counters } : {}),
        });
    }
    /**
     * This is a DEPLOY event, track some additional statistics about it
     *
     * We use this to measure things about deployment failures, such as the number of
     * failed DEPLOY events in a sequence.
     */
    async trackDeployStatistics(isSuccessful, counters) {
        await (0, telemetry_state_1.withTelemetryState)((state) => {
            const recentFailures = state.sequentialDeploymentFailures ?? 0;
            if (isSuccessful) {
                state.sequentialDeploymentFailures = 0;
            }
            else {
                state.sequentialDeploymentFailures = recentFailures + 1;
            }
            counters.sequentialDeploymentFailures = state.sequentialDeploymentFailures;
        });
    }
    get sessionInfo() {
        if (!this._sessionInfo) {
            throw new toolkit_lib_1.ToolkitError('SessionNotInitialized', 'Session Info not initialized. Call begin() first.');
        }
        return this._sessionInfo;
    }
}
exports.TelemetrySession = TelemetrySession;
function getState(error) {
    if (error) {
        return isAbortedError(error) ? 'ABORTED' : 'FAILED';
    }
    return 'SUCCEEDED';
}
function isAbortedError(error) {
    if (error?.name === 'ToolkitError' && error?.message?.includes(ABORTED_ERROR_MESSAGE)) {
        return true;
    }
    return false;
}
function mutable(x) {
    return x;
}
/**
 * Validates that the CDK_CLI_USERAGENT env var value matches
 * the expected format: `<name>/<version>/<mode>` where name is one of
 * VALID_USER_AGENTS and mode is either `sandbox` or `production`.
 */
function isValidWrapperUserAgent(value) {
    if (!value)
        return false;
    const parts = value.split('/');
    if (parts.length !== 3)
        return false;
    const [name, _version, mode] = parts;
    if (!VALID_USER_AGENTS.includes(name))
        return false;
    return mode === 'sandbox' || mode === 'production';
}
